### PR TITLE
Permit an array of tuple type in an is-type expression or in a pattern.

### LINKED ISSUE #24975: Arrays of tuples don't work with `is`
I was directed here from https://github.com/dotnet/csharplang/issues/1334

**Version Used**: 

VBCSCompiler.exe is 2.6.0.62329, not sure if that's the version you're interested in

**Steps to Reproduce**:

Compile this:

    using System;
    class A
    {
        void Test()
        {
            (object, object)[] a = new (object, object)[1]; // This line has no errors
            var b = a as (object, object)[]; // No errors
            if (a is (object, object)[]) // CS1525: Invalid expression term 'object', CS0443: Syntax error; value expected
            {
                Console.WriteLine("Hello");
            }
            if (a is (object, object)[] c) // CS1525: Invalid expression term 'object', CS0443: Syntax error; value expected
            {
                Console.WriteLine("World");
            }
        }
    }

**Expected Behavior**:
Successful compilation

**Actual Behavior**:
See embedded comments and also https://github.com/dotnet/csharplang/issues/1334

### PR DESCRIPTION
**Pull request by gafter**
Fixes #24975
Also move some suggestions from PROTOTYPE comments to the test plan.

@cston @agocke @dotnet/roslyn-compiler Please review this parsing change for pattern-matching.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 1]
```diff
@@ -345,7 +347,8 @@ private CSharpSyntaxNode ParseExpressionOrPattern(bool whenIsKeyword, Precedence
                 TypeSyntax type = null;
                 if ((SyntaxFacts.IsPredefinedType(tk) || tk == SyntaxKind.IdentifierToken) &&
                       // If it is a nameof, skip the 'if' and parse as an expression. 
-                      (this.CurrentToken.ContextualKind != SyntaxKind.NameOfKeyword || this.PeekToken(1).Kind != SyntaxKind.OpenParenToken))
+                      (this.CurrentToken.ContextualKind != SyntaxKind.NameOfKeyword || this.PeekToken(1).Kind != SyntaxKind.OpenParenToken) ||
+                    LooksLikeTupleArrayType())
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: The parentheses around the subexpressions before `LooksLikeTupleArrayType()` are slightly different from `ParseTypeOrPatternForIsOperatorCore`. Is that intentional? If not, should this entire `if` condition be extracted to a helper method?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs [hunk 2]
```diff
@@ -1452,429 +1452,5 @@ class Cat {}
                 );
             var comp = CompileAndVerify(compilation, expectedOutput: @"Fox Cat");
         }
-
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs]: Were these tests removed intentionally?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -5957,13 +5957,14 @@ private ScanTypeFlags ScanNonArrayType(ParseTypeMode mode, out SyntaxToken lastT
...
                 result = ScanTypeFlags.MustBeType;
             }
-            else if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken && mode != ParseTypeMode.DefinitePattern)
+            else if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
             {
                 lastTokenOfType = this.EatToken();
 
                 result = this.ScanTupleType(out lastTokenOfType);
-                if (result == ScanTypeFlags.NotType)
+                if (result == ScanTypeFlags.NotType || mode == ParseTypeMode.DefinitePattern && this.CurrentToken.Kind != SyntaxKind.OpenBracketToken)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: The name `ScanNonArrayType` makes it sound like the precondition for calling this method is that the type is not an array. Is that not true?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs (hunk 2 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs]: No. Looks like a merge gone bad. Will reinsert them.

---
In reply to: [181198412](https://github.com/dotnet/roslyn/pull/26088#discussion_r181198412) [](ancestors = 181198412)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 3 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: It is, but only trivially so. Every array type starts with a non-array type. This code scans that non-array type. It refuses to scan the array brackets and just leaves them as the next tokens on the input.

---
In reply to: [181215276](https://github.com/dotnet/roslyn/pull/26088#discussion_r181215276) [](ancestors = 181215276)

**Pull request merged**